x/exchange/keeper/market.go

```diff
@@ -0,0 +1,1303 @@
+}
+
+// CanCancelOrdersForMarket returns true if the provided admin bech32 address has permission to
+// cancel orders for a market. Also returns true if the provided address is the authority address.
+func (k Keeper) CanCancelOrdersForMarket(ctx sdk.Context, marketID uint32, admin string) bool {
+	return k.HasPermission(ctx, marketID, admin, exchange.Permission_cancel)
+}
+
+// CanWithdrawMarketFunds returns true if the provided admin bech32 address has permission to
+// withdraw funds from the given market's account. Also returns true if the provided address is the authority address.
+func (k Keeper) CanWithdrawMarketFunds(ctx sdk.Context, marketID uint32, admin string) bool {
+	return k.HasPermission(ctx, marketID, admin, exchange.Permission_withdraw)
+}
+
+// CanUpdateMarket returns true if the provided admin bech32 address has permission to
+// update market details and settings. Also returns true if the provided address is the authority address.
+func (k Keeper) CanUpdateMarket(ctx sdk.Context, marketID uint32, admin string) bool {
+	return k.HasPermission(ctx, marketID, admin, exchange.Permission_update)
+}
+
+// CanManagePermissions returns true if the provided admin bech32 address has permission to
+// manage user permissions for a given market. Also returns true if the provided address is the authority address.
+func (k Keeper) CanManagePermissions(ctx sdk.Context, marketID uint32, admin string) bool {
+	return k.HasPermission(ctx, marketID, admin, exchange.Permission_permissions)
+}
+
+// CanManageReqAttrs returns true if the provided admin bech32 address has permission to
+// manage required attributes for a given market. Also returns true if the provided address is the authority address.
+func (k Keeper) CanManageReqAttrs(ctx sdk.Context, marketID uint32, admin string) bool {
+	return k.HasPermission(ctx, marketID, admin, exchange.Permission_attributes)
+}
+
+// GetUserPermissions gets all permissions that have been granted to a user in a market.
+func (k Keeper) GetUserPermissions(ctx sdk.Context, marketID uint32, addr sdk.AccAddress) []exchange.Permission {
+	return getUserPermissions(k.getStore(ctx), marketID, addr)
+}
+
+// GetAccessGrants gets all the access grants for a market.
+func (k Keeper) GetAccessGrants(ctx sdk.Context, marketID uint32) []exchange.AccessGrant {
+	return getAccessGrants(k.getStore(ctx), marketID)
+}
+
+// UpdatePermissions updates users permissions in the store using the provided changes.
+// The caller is responsible for making sure this update should be allowed (e.g. by calling CanManagePermissions first).
+func (k Keeper) UpdatePermissions(ctx sdk.Context, msg *exchange.MsgMarketManagePermissionsRequest) error {
+	admin := sdk.MustAccAddressFromBech32(msg.Admin)
+	marketID := msg.MarketId
+	store := k.getStore(ctx)
+	var errs []error
+
+	for _, addrStr := range msg.RevokeAll {
+		addr := sdk.MustAccAddressFromBech32(addrStr)
+		perms := getUserPermissions(store, marketID, addr)
+		if len(perms) > 0 {
```

Comment:
Loop control statements like `continue` and `break` make reading and understanding code more difficult. I'm not saying they should never be used, heck, I even know of valid uses of `goto`, but it's good to try to avoid them.

That being said, I did flatten this a bit:
```
		if len(perms) == 0 {
			errs = append(errs, fmt.Errorf("account %s does not have any permissions for market %d", addrStr, marketID))
		}
		if len(errs) == 0 {
			revokeUserPermissions(store, marketID, addr)
		}
```